llpc/lower/llpcSpirvLowerConstImmediateStore.cpp

```diff
@@ -51,164 +52,200 @@ namespace Llpc {
 // @param func : Function to process
-void SpirvLowerConstImmediateStore::processAllocaInsts(Function *func) {
+bool SpirvLowerConstImmediateStore::processAllocaInsts(Function *func) {
   // NOTE: We only visit the entry block on the basis that SPIR-V translator puts all "alloca"
   // instructions there.
-  auto entryBlock = &func->front();
-  for (auto instIt = entryBlock->begin(), instItEnd = entryBlock->end(); instIt != instItEnd; ++instIt) {
-    auto inst = &*instIt;
-    if (auto allocaInst = dyn_cast<AllocaInst>(inst)) {
-      if (allocaInst->getAllocatedType()->isAggregateType()) {
-        // Got an "alloca" instruction of aggregate type.
-        auto storeInst = findSingleStore(allocaInst);
-        if (storeInst && isa<Constant>(storeInst->getValueOperand())) {
-          // Got an aggregate "alloca" with a single store to the whole type.
-          // Do the optimization.
-          convertAllocaToReadOnlyGlobal(storeInst);
-        }
-      }
+  bool changed = false;
+  SmallVector<AllocaInst *> candidates;
+  for (auto &inst : func->getEntryBlock()) {
+    if (auto allocaInst = dyn_cast<AllocaInst>(&inst)) {
+      if (allocaInst->getAllocatedType()->isAggregateType())
+        candidates.push_back(allocaInst);
     }
   }
+  for (auto *alloca : candidates) {
+    if (tryProcessAlloca(alloca))
+      changed |= true;
+  }
+  return changed;
 }
 
 // =====================================================================================================================
-// Finds the single "store" instruction storing to this pointer.
-//
-// Returns nullptr if no "store", multiple "store", or partial "store" instructions (store only part
-// of the memory) are found.
-//
-// NOTE: This is conservative in that it returns nullptr if the pointer escapes by being used in anything
-// other than "store" (as the pointer), "load" or "getelementptr" instruction.
+// For a single alloca, try to replace it by a constant global variable.
 //
 // @param allocaInst : The "alloca" instruction to process
-StoreInst *SpirvLowerConstImmediateStore::findSingleStore(AllocaInst *allocaInst) {
-  std::vector<Instruction *> pointers;
-  bool isOuterPointer = true;
-  StoreInst *storeInstFound = nullptr;
-  Instruction *pointer = allocaInst;
-  while (true) {
-    for (auto useIt = pointer->use_begin(), useItEnd = pointer->use_end(); useIt != useItEnd; ++useIt) {
-      auto user = cast<Instruction>(useIt->getUser());
+// @return true if the alloca was replaced
+bool SpirvLowerConstImmediateStore::tryProcessAlloca(AllocaInst *allocaInst) {
+  // LLVM IR allocas can have an "arrayness" where multiple elements of the allocated type are allocated at once.
+  // SPIR-V doesn't have this (because it only has OpVariable and not a "true" alloca), but let's guard against it
+  // anyway just in case.
+  auto *allocaArraySize = dyn_cast<ConstantInt>(allocaInst->getArraySize());
+  if (!allocaArraySize || allocaArraySize->getZExtValue() > 1)
+    return false;
+
```

Comment:
Thanks for adding this, maybe you can use the handy helper `isArrayAllocation()`.